### DISCUSSION TITLE
pygraphviz_layout with fixed positions

### DISCUSSION THREAD

**Discussion by steven-bioinfo** [networkx/networkx#7433] [Q&A] [answered]
Dear community,

I have a hierarchical network DiGraph with nodes already positioned using pygraphviz_layout (program dot).

I want to add new nodes (and edges) to this DiGraph. These new nodes need to be positioned, but if I run a new pygraphviz_layout, the previously positioned nodes have their positions modified, and I don't want that. I want to position only the new nodes (relative to the old nodes).

I've only found the spring_layout function to fix node positions, but the results are not hierarchical and the spring_layout position is not very compatible with the position given by dot.
Is it possible to do the same thing to fix node positions with pygraphviz_layout or another function to call the dot graphviz program or an other hierarchical program with fixed nodes option ?  

Thank you for your help,

Steven

**Comment by dschult** [comment, marked as answer]
I think you are actually looking for a GraphViz feature...  pygraphviz is a python interface to the GraphViz library.

[It looks like](https://observablehq.com/@magjac/placing-graphviz-nodes-in-fixed-positions) adding a `!` character to the existing pos string [as described in the graphviz docs](https://www.graphviz.org/docs/attrs/pos/).  Then the `neato` and `fdp` layout routines take the dot file and add positions to the nodes that don't have `!` in their pos string.  There are many other sources of info for how graphviz does layout that I'm not familiar with. That might be a direction to focus your search.  

NetworkX's spring_layout allows you to fix some nodes (as you mention above).
